source/service_test.go

```diff
@@ -0,0 +1,201 @@
+package source
+
+import (
+	"testing"
+
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+)
+
+// TestEndpoints tests that various services generate the correct endpoints.
+func TestEndpoints(t *testing.T) {
+	for _, tc := range []struct {
+		namespace   string
+		name        string
+		annotations map[string]string
+		lbs         []string
+		expected    []endpoint.Endpoint
+	}{
+		// Completely opted-out: no endpoints returned.
+		{
+			"testing",
+			"foo",
+			map[string]string{},
+			[]string{"1.2.3.4"},
+			[]endpoint.Endpoint{},
+		},
+		// Opt-in by setting desired hostname.
+		{
+			"testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/hostname": "foo.example.org",
+			},
+			[]string{"1.2.3.4"},
+			[]endpoint.Endpoint{
+				{DNSName: "foo.example.org", Target: "1.2.3.4"},
+			},
+		},
+		// Opt-in by setting desired hostname and this controller.
+		{
+			"testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/controller": "dns-controller",
+				"external-dns.kubernetes.io/hostname":   "foo.example.org",
+			},
+			[]string{"1.2.3.4"},
+			[]endpoint.Endpoint{
+				{DNSName: "foo.example.org", Target: "1.2.3.4"},
+			},
+		},
+		// Opt-out by setting a different controller.
+		{
+			"testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/controller": "some-other-tool",
+				"external-dns.kubernetes.io/hostname":   "foo.example.org",
+			},
+			[]string{"1.2.3.4"},
+			[]endpoint.Endpoint{},
+		},
+		// Make sure services are found in all namespaces.
+		{
+			"other-testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/hostname": "foo.example.org",
+			},
+			[]string{"1.2.3.4"},
+			[]endpoint.Endpoint{
+				{DNSName: "foo.example.org", Target: "1.2.3.4"},
+			},
+		},
+		// No external entrypoints lead to no endpoints.
+		{
+			"testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/hostname": "foo.example.org",
+			},
+			[]string{},
+			[]endpoint.Endpoint{},
+		},
+		// Multiple external entrypoints lead to multiple endpoints.
+		{
+			"testing",
+			"foo",
+			map[string]string{
+				"external-dns.kubernetes.io/hostname": "foo.example.org",
+			},
+			[]string{"1.2.3.4", "8.8.8.8"},
+			[]endpoint.Endpoint{
+				{DNSName: "foo.example.org", Target: "1.2.3.4"},
+				{DNSName: "foo.example.org", Target: "8.8.8.8"},
+			},
+		},
+	} {
```

Comment:
This looks like it could run in parallel. In that case you could use `t.Run(tc.msg, func(t *testing.T) {})`: https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks :)